### PR TITLE
Process ChanCloseInit in relayer loop (start command)

### PR DESCRIPTION
**Pull request by ancazamfir**
Closes: #560

## Description

- collect `CloseInitChannel` event (`collect_events()`) and handle it in `build_msg_from_event()`
- handle the race condition where client updates are sent on destination for heights higher than event height:
   - added a new function `adjust_events_height()` that checks:
       - if latest consensus height <= event_height (no change) or
       - if a client consensus of height `event_height + 1` exists (no adjustment) or
       - one at height `> event_height + 1` (change events height to latest client consensus height - 1)
   - the race condition can be reproduce this way:
       - start the realyer loop on an open channel (`hermes start ibc-0 ibc-1 -p transfer -c channel-0`)
       - from one terminal send tens of packets to `ibc-0` (`hermes tx raw ft-transfer ibc-0 ibc-1 transfer channel-0 9999 1000 -n 30`) 
       - from a different terminal send a `chan-close-init` to `ibc-1` (`hermes tx raw chan-close-init ibc-1 ibc-0 connection-0 transfer transfer -d channel-1 -s channel-0`)
       - before the `CloseInitChannel` event with some height `X` from `ibc-1` is processed, the `SendPacket` events processing results in `MsgTimeoutOnClose` packets to be sent to `ibc-0` prepended by a `MsgClientUpdate` at some height > `X`. This causes the subsequent `chan-close-init` event processing to fail.
    - there are probably many other ways to cause the race

Minor changes:
- added `timeout_on_close_packet` event decoder, with some awkwardness because `message.action` is `timeout_on_close_packet ` while the event name (prefix of tags) is `timeout_packet`. Needs more digging in cosmos-sdk, maybe ask them to make the changes.
- changed the channel event names (removed `Channel` suffix) 
 

______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Review by romac — COMMENTED** [review]
Looks good to me, as far as I can tell it does indeed implement the fix described above. Just pushed a commit with some minor stylistic changes.

I have only one remark aside from the one inlined below: as the event processing logic is getting more and more complex, it might eventually warrant a refactoring, perhaps even be rewritten as a state machine of some kind? Just some food for thought, this should be done in a separate PR anyway.

File: relayer/src/link.rs [hunk 1]
```diff
@@ -165,22 +167,64 @@ impl RelayPath {
         event: &IBCEvent,
     ) -> Result<(Option<Any>, Option<Any>), LinkError> {
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/link.rs:168]: I believe we will need quite a bit of restructuring when we fix https://github.com/informalsystems/ibc-rs/issues/640. Would also require some sort of packet sending thread to implement a delay period timer and send packets when the timer fires. Let's do these changes there, including the refactoring.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,55 @@
 
 [comment]: <> (## Unreleased Changes)
```
Comment (romac) [inline CHANGELOG.md:3]: This should be uncommented now that there is content for unreleased changes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,55 @@
...
 
+### FEATURES
+
+- <high level summary>
+
+- [modules]
```
Comment (romac) [inline CHANGELOG.md:9]: We should use the actual crate names instead of the folder names, ie. ibc, ibc-relayer, ibc-relayer-cli

**Review by romac — CHANGES_REQUESTED** [review]
Can we also fix or silence the clippy warnings, depending on whether we expect those functions to potentially returns errors at some point?

**Review by romac — APPROVED** [review]

**Pull request merged**